src/passes/MultiMemoryLowering.cpp

```diff
@@ -0,0 +1,431 @@
+        addGlobal(name, offsetRunningTotal * Memory::kPageSize);
+      }
+      offsetRunningTotal += memory->initial;
+    }
+  }
+
+  void adjustActiveDataSegmentOffsets() {
+    Builder builder(*wasm);
+    ModuleUtils::iterActiveDataSegments(*wasm, [&](DataSegment* dataSegment) {
+      assert(dataSegment->offset->is<Const>() &&
+             "TODO: handle non-const segment offsets");
+      auto idx = memoryIdxMap.at(dataSegment->memory);
+      dataSegment->memory = combinedMemory;
+      // No need to update the offset of data segments for the first memory
+      if (idx != 0) {
+        auto offsetGlobalName = getOffsetGlobal(idx);
+        assert(wasm->features.hasExtendedConst());
+        dataSegment->offset = builder.makeBinary(
+          Abstract::getBinary(pointerType, Abstract::Add),
+          builder.makeGlobalGet(offsetGlobalName, pointerType),
+          dataSegment->offset);
+      }
+    });
+  }
+
+  void createMemorySizeFunctions() {
+    for (Index i = 0; i < wasm->memories.size(); i++) {
+      auto function = memorySize(i, wasm->memories[i]->name);
+      memorySizeNames.push_back(function->name);
+      wasm->addFunction(std::move(function));
+    }
+  }
+
+  void createMemoryGrowFunctions() {
+    for (Index i = 0; i < wasm->memories.size(); i++) {
+      auto function = memoryGrow(i, wasm->memories[i]->name);
+      memoryGrowNames.push_back(function->name);
+      wasm->addFunction(std::move(function));
+    }
+  }
+
+  // This function replaces memory.grow instruction calls in the wasm module.
+  // Because the multiple discrete memories are lowered into a single memory,
+  // we need to adjust offsets as a particular memory receives an
+  // instruction to grow.
+  std::unique_ptr<Function> memoryGrow(Index memIdx, Name memoryName) {
+    Builder builder(*wasm);
+    Name name = memoryName.toString() + "_grow";
+    Name functionName = Names::getValidFunctionName(*wasm, name);
+    auto function = Builder::makeFunction(
+      functionName, Signature(pointerType, pointerType), {});
+    function->setLocalName(0, "page_delta");
+    Index pageSizeLocal = -1;
+    auto pageSizeConst = [&]() {
+      return builder.makeLocalGet(pageSizeLocal, Type::i32);
+    };
+    auto getOffsetDelta = [&]() {
+      return builder.makeBinary(Abstract::getBinary(pointerType, Abstract::Mul),
+                                builder.makeLocalGet(0, pointerType),
+                                pageSizeConst());
+    };
+    auto getMoveSource = [&](Name global) {
+      return builder.makeGlobalGet(global, pointerType);
+    };
+    Expression* functionBody;
+    Index sizeLocal = -1;
+
+    Index returnLocal =
+      Builder::addVar(function.get(), "return_size", pointerType);
+    functionBody = builder.blockify(builder.makeLocalSet(
+      returnLocal, builder.makeCall(memorySizeNames[memIdx], {}, pointerType)));
+
+    if (!isLastMemory(memIdx)) {
+      pageSizeLocal = Builder::addVar(function.get(), "page_size", Type::i32);
```

Comment:
A page size local can be avoided, as the `makeConst` can just be done wherever the `local.get` of the local is atm. (There isn't much cost to having multiple appearances of it, too, as it is just a constant. Other things might be different if they take work at runtime though, like an add.)